File: src/agent_dump/selector.py
"""
Session selection utilities
"""

from datetime import datetime, timedelta
import sys

import questionary
from questionary import Choice, Style

from agent_dump.agents.base import BaseAgent, Session
from agent_dump.i18n import Keys, i18n
from agent_dump.rendering import format_session_metadata_summary
from agent_dump.text_safety import safe_display_text
from agent_dump.time_utils import get_local_timezone, to_local_datetime


def is_terminal() -> bool:
    """Check if running in a terminal"""
    return sys.stdin.isatty() and sys.stdout.isatty()


def get_time_group(session: Session) -> str:
    """Get time group for a session"""
    local_tz = get_local_timezone()
    now = datetime.now(local_tz)
    today = now.replace(hour=0, minute=0, second=0, microsecond=0)
    yesterday = today - timedelta(days=1)
    week_ago = today - timedelta(days=7)
    month_ago = today - timedelta(days=30)

    session_time = to_local_datetime(session.created_at, local_tz)

    if session_time >= today:
        return i18n.t(Keys.TIME_TODAY)
    elif session_time >= yesterday:
        return i18n.t(Keys.TIME_YESTERDAY)
    elif session_time >= week_ago:
        return i18n.t(Keys.TIME_THIS_WEEK)
    elif session_time >= month_ago:
        return i18n.t(Keys.TIME_THIS_MONTH)
    else:
        return i18n.t(Keys.TIME_OLDER)


def group_sessions(sessions: list[Session]) -> dict[str, list[Session]]:
    """Group sessions by time periods"""
    groups: dict[str, list[Session]] = {}

    for session in sessions:
        group = get_time_group(session)
        if group not in groups:
            groups[group] = []
        groups[group].append(session)

    # Define order
    order = [
        i18n.t(Keys.TIME_TODAY),
        i18n.t(Keys.TIME_YESTERDAY),
        i18n.t(Keys.TIME_THIS_WEEK),
        i18n.t(Keys.TIME_THIS_MONTH),
        i18n.t(Keys.TIME_OLDER),
        i18n.t(Keys.TIME_UNKNOWN),
    ]
    ordered_groups = {}
    for key in order:
        if key in groups:
            ordered_groups[key] = groups[key]

    return ordered_groups


def select_agent_interactive(agents: list[BaseAgent], session_counts: dict[str, int]) -> BaseAgent | None:
    """Let user select an agent tool interactively.

    session_counts 是必填的：selector 属于 UI 层，不得自己去扫 provider
    （AGENTS.md §1.4）。调用方本来就要扫一次，把计数传进来即可。
    """
    if not agents:
        print(i18n.t(Keys.NO_AGENTS_FOUND))
        return None

    if not is_terminal():
        return select_agent_simple(agents, session_counts)

    choices = []
    for agent in agents:
        count = session_counts.get(agent.name, 0)
        label = f"{agent.display_name} ({count} {i18n.t(Keys.SESSION_COUNT_SUFFIX)})"
        choices.append(questionary.Choice(title=label, value=agent))

    custom_style = Style(
        [
            ("qmark", "fg:#673ab7 bold"),
            ("question", "bold"),
            ("answer", "fg:#f44336 bold"),
            ("pointer", "fg:#673ab7 bold"),
            ("highlighted", "noreverse"),
            ("selected", "noreverse"),
            ("separator", "fg:#cc5454"),
            ("instruction", ""),
            ("text", ""),
        ]
    )

    q = questionary.select(
        i18n.t(Keys.SELECT_AGENT_PROMPT),
        choices=choices,
        style=custom_style,
        instruction=i18n.t(Keys.SELECT_INSTRUCTION),
    )

    if q.application.key_bindings:
        q.application.key_bindings.add("q")(lambda event: event.app.exit(result=None))  # type: ignore
        q.application.key_bindings.add("Q")(lambda event: event.app.exit(result=None))  # type: ignore

    try:
        selected = q.ask()
    except KeyboardInterrupt:
        print("\n" + i18n.t(Keys.USER_CANCELLED))
        return None

    return selected


def select_agent_simple(agents: list[BaseAgent], session_counts: dict[str, int]) -> BaseAgent | None:
    """Simple agent selection for non-terminal environments"""
    print(i18n.t(Keys.AVAILABLE_AGENTS))
    print("-" * 80)
    for i, agent in enumerate(agents, 1):
        count = session_counts.get(agent.name, 0)
        print(f"{i}. {agent.display_name} ({count} {i18n.t(Keys.SESSION_COUNT_SUFFIX)})")
    print()

    print(i18n.t(Keys.SELECT_AGENT_NUMBER))
    try:
        selection = input("> ").strip()
    except EOFError:
        print("\n" + i18n.t(Keys.NO_INPUT_EXITING))
        return None

    try:
        idx = int(selection) - 1
        if 0 <= idx < len(agents):
            return agents[idx]
        else:
            print(i18n.t(Keys.INVALID_SELECTION, selection=selection))
            return None
    except ValueError:
        print(i18n.t(Keys.INVALID_INPUT_NUMBER))
        return None


def select_sessions_interactive(
    sessions: list[Session], agent: BaseAgent, show_metadata_summary: bool = True
) -> list[Session]:
    """Let user select sessions interactively with time grouping"""
    if not sessions:
        print(i18n.t(Keys.NO_SESSIONS_IN_RANGE))
        return []

    if not is_terminal():
        return select_sessions_simple(sessions, agent, show_metadata_summary=show_metadata_summary)

    # Group sessions by time
    groups = group_sessions(sessions)

    choices = []

    for group_name, group_sessions_list in groups.items():
        # Add separator for group
        choices.append(
            Choice(
                title=i18n.t(Keys.GROUP_TITLE, group_name=group_name, count=len(group_sessions_list)),
                disabled=i18n.t(Keys.GROUP_HEADER_DISABLED_HINT),
            )
        )

        # Add sessions in this group
        for session in group_sessions_list:
            # 标题、summary、URI 全部来自第三方 Session；先逐个净化成单行，再由这里
            # 加入有意的换行。对拼好的整块调 safe_display_text 会把布局一起压平。
            label = safe_display_text(agent.get_formatted_title(session))
            if show_metadata_summary:
                summary = safe_display_text(format_session_metadata_summary(agent, session))
                title = f"  {label}\n    {summary}"
            else:
                uri = safe_display_text(agent.get_session_uri(session))
                title = f"  {label} {uri}"
            choices.append(questionary.Choice(title=title, value=session))

    custom_style = Style(
        [
            ("qmark", "fg:#673ab7 bold"),
            ("question", "bold"),
            ("answer", "fg:#f44336 bold"),
            ("pointer", "fg:#673ab7 bold"),
            ("highlighted", "noreverse"),
            ("selected", "noreverse"),
            ("separator", "fg:#cc5454"),
            ("instruction", ""),
            ("text", ""),
            ("disabled", "fg:#666666 italic"),
        ]
    )

    q = questionary.checkbox(
        i18n.t(Keys.SELECT_SESSIONS_PROMPT),
        choices=choices,
        style=custom_style,
        instruction=i18n.t(Keys.CHECKBOX_INSTRUCTION),
    )

    if q.application.key_bindings:
        q.application.key_bindings.add("q")(lambda event: event.app.exit(result=None))  # type: ignore
        q.application.key_bindings.add("Q")(lambda event: event.app.exit(result=None))  # type: ignore

    try:
        selected = q.ask()
    except KeyboardInterrupt:
        print("\n" + i18n.t(Keys.USER_CANCELLED))
        return []

    return selected or []


def select_sessions_simple(
    sessions: list[Session], agent: BaseAgent, show_metadata_summary: bool = True
) -> list[Session]:
    """Simple selection for non-terminal environments"""
    # Group sessions for display
    groups = group_sessions(sessions)

    print(i18n.t(Keys.AVAILABLE_SESSIONS))
    print("-" * 80)

    idx = 1
    session_map = {}

    for group_name, group_sessions_list in groups.items():
        print(f"\n[{group_name}] ({len(group_sessions_list)} {i18n.t(Keys.SESSION_COUNT_SUFFIX)})")
        for session in group_sessions_list:
            title = safe_display_text(agent.get_formatted_title(session))
            if show_metadata_summary:
                print(f"{idx}. {title}")
                print(f"    {safe_display_text(format_session_metadata_summary(agent, session))}")
            else:
                uri = safe_display_text(agent.get_session_uri(session))
                print(f"{idx}. {title} {uri}")
            session_map[idx] = session
            idx += 1

    print()
    print(i18n.t(Keys.ENTER_SESSION_NUMBERS))
    try:
        selection = input("> ").strip()
    except EOFError:
        print("\n" + i18n.t(Keys.NO_INPUT_EXITING))
        return []

    if selection.lower() == "all":
        return sessions

    try:
        indices = [int(x.strip()) for x in selection.split(",")]
        selected = []
        for num in indices:
            if num in session_map:
                selected.append(session_map[num])
            else:
                print(i18n.t(Keys.INVALID_SELECTION, selection=num))
        return selected
    except ValueError:
        print(i18n.t(Keys.INVALID_INPUT_NUMBERS))
        return []

File: rust/src/selector.rs
use crate::i18n::{t, terminal};
use crate::session::Session;
use crate::tui::Row;
use std::io::{BufRead, Write};

fn input(
    out: &mut impl Write,
    reader: &mut impl BufRead,
    zh: bool,
) -> crate::Result<Option<String>> {
    write!(out, "> ")?;
    out.flush()?;
    let mut text = String::new();
    if reader.read_line(&mut text)? == 0 {
        writeln!(out, "\n{}", t("NO_INPUT_EXITING", zh, &[]))?;
        return Ok(None);
    }
    Ok(Some(text.trim().into()))
}

pub fn provider(
    rows: &[Row],
    zh: bool,
    out: &mut impl Write,
    reader: &mut impl BufRead,
) -> crate::Result<Option<usize>> {
    if crate::tui::available() {
        out.flush()?;
        return Ok(
            crate::tui::select(&t("SELECT_AGENT_PROMPT", zh, &[]), rows, false, 0, zh)?
                .and_then(|v| v.first().copied()),
        );
    }
    writeln!(
        out,
        "{}\n{}",
        t("AVAILABLE_AGENTS", zh, &[]),
        "-".repeat(80)
    )?;
    for (index, row) in rows.iter().enumerate() {
        writeln!(out, "{}. {}", index + 1, row.title)?;
    }
    writeln!(out, "\n{}", t("SELECT_AGENT_NUMBER", zh, &[]))?;
    let Some(text) = input(out, reader, zh)? else {
        return Ok(None);
    };
    match integer(&text) {
        Some(index) if index > 0 && index as usize <= rows.len() => Ok(Some(index as usize - 1)),
        Some(_) => {
            writeln!(
                out,
                "{}",
                terminal("INVALID_SELECTION", zh, &[("selection", text)])
            )?;
            Ok(None)
        }
        None => {
            writeln!(out, "{}", t("INVALID_INPUT_NUMBER", zh, &[]))?;
            Ok(None)
        }
    }
}

fn integer(value: &str) -> Option<i128> {
    let normalized: String = value.chars().map(crate::value::decimal_digit).collect();
    if normalized.contains("__")
        || normalized.starts_with('_')
        || normalized.ends_with('_')
        || normalized.starts_with("+_")
        || normalized.starts_with("-_")
    {
        return None;
    }
    normalized.replace('_', "").parse().ok()
}

pub fn sessions(
    sessions: &[Session],
    scheme: &str,
    summary: bool,
    zh: bool,
    out: &mut impl Write,
    reader: &mut impl BufRead,
) -> crate::Result<Vec<usize>> {
    let today = jiff::Zoned::now().date();
    let keys = [
        "TIME_TODAY",
        "TIME_YESTERDAY",
        "TIME_THIS_WEEK",
        "TIME_THIS_MONTH",
        "TIME_OLDER",
    ];
    let mut groups: [Vec<usize>; 5] = Default::default();
    for (index, session) in sessions.iter().enumerate() {
        let days = today
            .since((jiff::Unit::Day, session.created_at.local_date()))?
            .get_days();
        let group = match days {
            ..=0 => 0,
            1 => 1,
            2..=7 => 2,
            8..=30 => 3,
            _ => 4,
        };
        groups[group].push(index);
    }
    let mut rows = Vec::new();
    let mut map = Vec::new();
    for (group, indices) in groups.iter().enumerate().filter(|(_, v)| !v.is_empty()) {
        let group_title = t(keys[group], zh, &[]);
        for (position, &index) in indices.iter().enumerate() {
            let session = &sessions[index];
            let mut title = crate::render::formatted_title(session);
            let detail = if summary {
                crate::render::metadata_summary(session, scheme, zh)
            } else {
                title += &format!(
                    " {}",
                    crate::render::safe_line(&format!("{scheme}://{}", session.id))
                );
                String::new()
            };
            rows.push(Row {
                title,
                detail,
                group: if position == 0 {
                    format!(
                        "[{group_title}] ({} {})",
                        indices.len(),
                        t("SESSION_COUNT_SUFFIX", zh, &[])
                    )
                } else {
                    String::new()
                },
            });
            map.push(index);
        }
    }
    if crate::tui::available() {
        out.flush()?;
        return Ok(
            crate::tui::select(&t("SELECT_SESSIONS_PROMPT", zh, &[]), &rows, true, 0, zh)?
                .unwrap_or_default()
                .iter()
                .map(|i| map[*i])
                .collect(),
        );
    }
    writeln!(
        out,
        "{}\n{}",
        t("AVAILABLE_SESSIONS", zh, &[]),
        "-".repeat(80)
    )?;
    for (index, row) in rows.iter().enumerate() {
        if !row.group.is_empty() {
            writeln!(out, "\n{}", row.group)?;
        }
        writeln!(out, "{}. {}", index + 1, row.title)?;
        if !row.detail.is_empty() {
            writeln!(out, "    {}", row.detail)?;
        }
    }
    writeln!(out, "\n{}", t("ENTER_SESSION_NUMBERS", zh, &[]))?;
    let Some(text) = input(out, reader, zh)? else {
        return Ok(Vec::new());
    };
    if text.eq_ignore_ascii_case("all") {
        return Ok((0..sessions.len()).collect());
    }
    let Some(indices) = text
        .split(',')
        .map(|v| integer(v.trim()))
        .collect::<Option<Vec<_>>>()
    else {
        writeln!(out, "{}", t("INVALID_INPUT_NUMBERS", zh, &[]))?;
        return Ok(Vec::new());
    };
    let mut selected = Vec::new();
    for index in indices {
        if index > 0 && (index as usize) <= map.len() {
            selected.push(map[index as usize - 1]);
        } else {
            writeln!(
                out,
                "{}",
                t("INVALID_SELECTION", zh, &[("selection", index.to_string())])
            )?;
        }
    }
    Ok(selected)
}
